server/graphql/v1/mutations/orders.js

```diff
@@ -158,20 +158,69 @@ const checkGuestContribution = async (order, loaders) => {
     throw new BadRequest('You need to provide a guest profile with an email for logged out contributions');
   } else if (!guestInfo.email || !isEmail(guestInfo.email)) {
     throw new BadRequest('You need to provide a valid email');
-  } else if (order.totalAmount > 25000) {
-    const location = guestInfo.location || {};
-    if (!guestInfo.name) {
-      throw new BadRequest('Contributions that are more than $250 must have a name attached');
-    } else if (order.totalAmount > 500000 && !location.structured && (!location.address || !location.country)) {
-      throw new BadRequest('Contributions that are more than $5000 must have an address attached');
-    }
   } else if (order.fromCollective) {
     throw new BadRequest('You need to be logged in to specify a contributing profile');
   } else if (order.paymentMethod?.id || order.paymentMethod?.uuid) {
     throw new BadRequest('You need to be logged in to be able to use an existing payment method');
   }
 };
 
+const mustUpdateLocation = (existingLocation, newLocation) => {
+  const simpleFields = ['country', 'name', 'address'];
+  const hasUpdatedSimpleField = field => newLocation[field] && newLocation[field] !== existingLocation[field];
+  return simpleFields.some(hasUpdatedSimpleField) || !isEqual(existingLocation.structured, newLocation.structured);
+};
+
+const mustUpdateNames = (fromAccount, fromAccountInfo) => {
+  return (!fromAccount.name && fromAccountInfo?.name) || (!fromAccount.legalName && fromAccountInfo?.legalName);
+};
+
+/**
+ * Checks that the profile has all requirements for this contribution (name, address, etc) and updates
+ * it if necessary. Must be called **after** checking permissions and authentication!
+ */
+const checkAndUpdateProfileInfo = async (order, fromAccount, isGuest) => {
+  const { totalAmount, fromAccountInfo, guestInfo } = order;
+  const accountUpdatePayload = {};
+  const location = fromAccountInfo?.location || guestInfo?.location || fromAccount.location;
+
+  // Only enforce profile checks for guests at the moment
+  if (isGuest) {
+    // Contributions that are more than $5000 must have an address attached
+    if (totalAmount > 5000e2) {
+      if (!location.structured && (!location.address || !location.country)) {
+        throw new BadRequest('Contributions that are more than $5000 must have an address attached');
+      }
+    }
+
+    // Contributions that are more than $250 must have a name attached
+    if (totalAmount > 250e2) {
+      const name = fromAccountInfo?.name || fromAccountInfo?.legalName || fromAccount.name || fromAccount.legalName;
+      if (!name) {
+        throw new BadRequest('Contributions that are more than $250 must have a name attached');
+      }
+    }
+  }
+
+  // Update account with new info, unless we're making a guest contribution for an existing account
+  // (we don't want to let guests update the profile of an existing account that they may not own)
+  if (!isGuest || fromAccount.data?.isGuest) {
```

Comment:
This conditional is a bit cryptic and looks broken.
Are we missing an `!` before `fromAccount.data.isGuest`?
Maybe we declare an auxiliary variable like:

```js
const userIsExistingGuest = !!fromAccount.data?.isGuest;
if (!isGuest || userIsExistingGuest) {
```

Reply:
It's not broken, but it's a bit hard to parse indeed. The idea is:
- Either you're not contributing as a guest, and it's fine to update your profile directly (because you're authenticated)
- Otherwise, you can only update the profile if it's not verified yet (because we don't want anyone submitting an order with your email to be able to edit your profile)

Your suggestion looks good, I'm gonna do something in this spirit

```suggestion
  const existingProfileIsGuest = Boolean(fromAccount.data?.isGuest);
  if (!isGuest || existingProfileIsGuest) {
```